text/2021-08-11-implement-physical-machine-chaos.md

```diff
@@ -0,0 +1,143 @@
+tool, and also can work as a server. It lacks UI（just like Chaos Mesh Dashboard
+to create and manage experiments, and can't orchestrate the experiment just like
+the [workflow](https://chaos-mesh.org/docs/create-chaos-mesh-workflow).
+
+## Proposal
+
+Implement physical machine chaos in Chaos Mesh, and then we can reuse the
+Dashboard and workflow for it.
+
+There are two ways to implement it:
+
+### 1. Treat physical machines as a new selector
+
+In this way, we will reuse the chaos type in Chaos Mesh, and implement a new
+selector to choose physical machines.
+
+For example, below is a YAML config for NetworkChaos:
+
+```YAML
+apiVersion: chaos-mesh.org/v1alpha1
+kind: NetworkChaos
+metadata:
+  name: network-delay
+  namespace: busybox
+spec:
+  action: delay
+  mode: all
+  selector:
+    namespaces:
+      - busybox
+  delay:
+    latency: "5s"
+```
+
+It will inject network delay on all pods in the busybox namespace.
+
+For physical machine, we can implement a new selector, the YAML config may looks
+like:
+
+```YAML
+apiVersion: chaos-mesh.org/v1alpha1
+kind: NetworkChaos
+metadata:
+  name: network-delay
+  namespace: chaos-testing
+spec:
+  action: delay
+  selector:
+    physicalMachines:
+      - 123.123.123.123:123
+      - 124.124.124.124:124
+  delay:
+    latency: "5s"
+```
+
+We replace the selector `namespaces` to `physicalMachines` , and the
+`123.123.123.123:123` and `124.124.124.124:124` are the addresses of Chaosd server.
+
+#### Advantage of proposal 1
+
+The physics experiment and K8s experiment are unified, the only difference is
+the selectors.
+
+#### Disadvantage of proposal 1
+
+Higher implementation costs.
+
+* We know that Chaos Mesh is designed for K8s, there are too many codes coupled
+to K8s.
+* The chaos type is related to a selector, which means each chaos has only one
+target type. For example, DNSChaos inject fault to some containers, NetworkChaos
+inject fault to some pods. This means we need to implement the physical machine
+selector for every chaos.
+* The config and implementation of the same chaos type for K8s and physical
+machines are different, like DNSChaos, JVMChaos, etc. It is difficult to unify them.
+
+### 2. Treat chaos on the physical machine as a new chaos type
+
+Implement physical machine chaos as a new chaos type in Chaos Mesh. Add a new
+crd named `PhysicalMachineChaos` , the config includes:
+
+* action: the subtypes of `PhysicalMachineChaos`, the action can be `stress-cpu`,
+`stress-mem`, `network-delay`, `network-loss` and so on.
+* address: the addresses of chaosd server.
+* config related to the action: for example, for `stress-cpu` action, we need to
+set `load` and `workers`.
+
+Here is the sample YAML config for network delay:
+
+```YAML
+apiVersion: chaos-mesh.org/v1alpha1
+kind: PhysicalMachineChaos
+metadata:
+  name: physical-network-delay
+  namespace: chaos-testing
+spec:
+  action: network-delay
+  address:
+    - http://172.16.112.130:31767
```

Comment:
I also prefer the second option, but I have some questions here:

1. How to specify the target chaos daemons by this `PhysicalMachineChaos`?
 
After introducing this resource, we may need some separate PhysicalMachineChaos selectors at the next step.

2. How to manage these chaos daemons?

We did have a [daemonset](https://github.com/chaos-mesh/chaos-mesh/blob/master/helm/chaos-mesh/templates/chaos-daemon-daemonset.yaml) for controlling the daemons. If we introduce the `PhysicalMachineChaos`, those daemons weren't managed.

I'm unsure if it's needed to create sth like a chaos-daemon controller and maintain the daemons' state, once the chaos daemon started, it would register as a `ChaosNode` resource in k8s, and chaos-daemon controller would maintain its status, just like [k8s node controller did](https://kubernetes.io/docs/concepts/architecture/nodes/#node-controller). As an alternative, we could also just keep them being out of the cluster. After all, it's not a big problem.

BTW, it's so cool to support directly run experiments on vms / bms ! With the help of workflow and `PhysicalMachineChaos`, now we could orchestra a full experiment with in-cluster and out-cluster instances!

Reply:
1. We will add a new API to register chaosd's address with a meaningful name in Dashboard, and when create a new physical machine chaos, users can select the chaosd's address according to the name.

2.  Yeah, maybe need to add a heartbeat mechanism for chaosd, chaosd will report its state to controller-manager periodically. As the `ChaosNode` resource, it sounds cool! I will consider this way.

Thanks!

Reply:
> users can select the chaosd's address according to the name.

It would be a bonus feature if we support selecting the chaos daemons by label selectors 😁.

Therefore, our users could use an iaas-level solution, e.g. Ansible, to deploy chaos daemons with their own labels, and chaos controller could set targets in `PhysicalMachineChaos` by labels, which is similar to the Pod selectors!

Reply:
So controller-manager need to know the labels set by Ansible, maybe we can integrate a similar solution in controller-manager, just like [tiup](https://docs.pingcap.com/tidb/stable/production-deployment-using-tiup) 🤔